src/node_api.cc

```diff
@@ -3554,3 +3557,299 @@ napi_status napi_run_script(napi_env env,
+
+    node::CallbackScope(ts_fn->env->isolate,
+        v8::Local<v8::Object>::Cast(v8impl::V8LocalValueFromJsValue(js_cb)),
+        {0, 0});
+
+    ts_fn->call_js_cb(ts_fn->env, js_cb, ts_fn->hint, data);
+  }
+
+  if (size == 0) {
+    if (uv_idle_stop(&ts_fn->idle) != 0) {
+      napi_throw_error(ts_fn->env, nullptr,
+          "N-API thread-safe function: Failed to stop idle loop");
+      return;
+    }
+    ts_fn->idle_running = false;
+  }
+
+  if (delete_function) {
+    v8impl::Reference::Delete(ts_fn->ref);
+    uv_mutex_destroy(&ts_fn->mutex);
+
+    if (ts_fn->max_queue_size > 0) {
+      uv_cond_destroy(&ts_fn->cond);
+    }
+
+    // It is important that we delete the async first, because the async
+    // callback is the only place from which we awaken the idler. We must avoid
+    // waking the idler because we will have partially destroyed the thread-safe
+    // function by the time the idle handler runs. After all, we have already
+    // destroyed the mutex, the reference, and potentially the condition in the
+    // code above.
+    uv_close(reinterpret_cast<uv_handle_t*>(&ts_fn->async),
+        [] (uv_handle_t* handle) -> void {
+          napi_threadsafe_function ts_fn =
+              node::ContainerOf(&napi_threadsafe_function__::async,
+                reinterpret_cast<uv_async_t*>(handle));
+          uv_close(reinterpret_cast<uv_handle_t*>(&ts_fn->idle),
+            [] (uv_handle_t* handle) -> void {
+              napi_threadsafe_function ts_fn =
+                  node::ContainerOf(&napi_threadsafe_function__::idle,
+                    reinterpret_cast<uv_idle_t*>(handle));
+              if (ts_fn->thread_finalize_callback) {
+                v8::HandleScope handle_scope(ts_fn->env->isolate);
+                node::CallbackScope(ts_fn->env->isolate,
+                    v8::Object::New(ts_fn->env->isolate), {0, 0});
+                ts_fn->thread_finalize_callback(ts_fn->env,
+                    ts_fn->thread_finalize_data, ts_fn->hint);
+              }
+              delete ts_fn;
+            });
+        });
+  }
+}
+
+static void napi_threadsafe_function_cb(uv_async_t* uv_async) {
+  napi_threadsafe_function ts_fn =
+      node::ContainerOf(&napi_threadsafe_function__::async, uv_async);
+
+  v8::HandleScope handle_scope(ts_fn->env->isolate);
+
+  if (!ts_fn->idle_running) {
+    if (uv_idle_start(&ts_fn->idle, napi_threadsafe_function_idle_cb) != 0) {
+      napi_throw_error(ts_fn->env, nullptr,
+        "N-API thread-safe function: Failed to start idle loop");
+      return;
+    }
+    ts_fn->idle_running = true;
+  }
+}
+
+static void napi_threadsafe_function_default_call_js(napi_env env,
+                                                     napi_value js_cb,
+                                                     void* hint,
+                                                     void* data) {
+  napi_value recv;
+  napi_status status;
+  status = napi_get_global(env, &recv);
```

Comment:
Can you explain why we are using the global object for this? After all, JS strict mode very deliberately switched to not providing a default receiver…